components/usage/pkg/controller/reconciler.go

```diff
@@ -71,11 +74,59 @@ func (u *UsageReconciler) ReconcileTimeRange(ctx context.Context, from, to time.
 	if len(invalidInstances) > 0 {
 		log.WithField("invalid_workspace_instances", invalidInstances).Errorf("Detected %d invalid instances. These will be skipped in the current run.", len(invalidInstances))
 	}
-
 	log.WithField("workspace_instances", instances).Debug("Successfully loaded workspace instances.")
+
+	workspaces, err := u.loadWorkspaces(ctx, instances)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load workspaces for workspace instances in time range: %w", err)
+	}
+	status.Workspaces = len(workspaces)
```

Comment:
So, at this point we have:
- a collection of all "fully hydrated" instances
- a collection of all "fully hydrated" workspaces, each with an attached sub-collection of all its corresponding "fully hydrated" instances (although these will just be references to a single instance in memory, right, not a full copy of the object?)

I'm still slightly uneasy with loading so many large objects into memory, and iterating over them multiple times. So, I'm looking forward to seeing exactly how we'll use these collections, and see if we can later reduce the amount of used memory and the number of iterations. 👀